runners: probe-rs: add support to flash hex file

AndreasErnst-EH opened the pull request:
The probe-rs runner always flashed the elf file. With this commit support for the --hex-file parameter is added. This can be helpful when wanting to flash a signed binary instead of the unsigned elf file.

tpambor commented on scripts/west_commands/runners/probe_rs.py:
```diff
@@ -32,6 +32,10 @@ def __init__(self, cfg, chip,
             self.args += tool_opt
 
         self.elf_name = cfg.elf_file
+        if cfg.hex_file is not None:
+            self.hex_file = cfg.hex_file
+        else:
+            self.hex_file = None
```
The `if` can be omitted. Most runners call this `hex_name`, so I would suggest naming this here `hex_name` as well for consistency.
```suggestion
        self.hex_name = cfg.hex_file
```

tpambor commented on scripts/west_commands/runners/probe_rs.py:
```diff
@@ -88,7 +92,12 @@ def do_flash(self, **kwargs):
+            download_args += ['--binary-format', 'hex']
+            download_args += [self.hex_file]
+        else:
+            download_args += [self.elf_name]
```
If nothing is specified, probe-rs uses the preferred format of the target, this might happen to be elf. Let's specify elf here explicitly to be on the safe side.

```suggestion
            download_args += ['--binary-format', 'elf']
            download_args += [self.elf_name]
```

AndreasErnst-EH commented on scripts/west_commands/runners/probe_rs.py:
```diff
@@ -88,7 +89,13 @@ def do_flash(self, **kwargs):
+            download_args += ['--binary-format', 'hex']
+            download_args += [self.hex_file]
+        else:
+            download_args += ['--binary-format', 'elf']
```
Checked the behavior and when manually removing the elf_file entry from the runners.yaml then the behavior can be triggered. 
Added the Copilot recommendation with a slightly adapted error message which conforms more to the error message in other runners.

pdgendt commented on scripts/west_commands/runners/probe_rs.py:
```diff
@@ -88,7 +89,15 @@ def do_flash(self, **kwargs):
+        if self.hex_file is not None:
+            download_args += ['--binary-format', 'hex']
+            download_args += [self.hex_file]
+        else:
+            download_args += ['--binary-format', 'elf']
+            download_args += [self.elf_name]
+            if self.elf_name is None:
+                raise ValueError(f'Cannot flash; elf file required.')
```
Move this check to right after the `else` line?

```suggestion
        else:
            if self.elf_name is None:
                raise ValueError(f'Cannot flash; elf file required.')
            download_args += ['--binary-format', 'elf']
            download_args += [self.elf_name]
```

marc-hb commented on scripts/west_commands/runners/probe_rs.py:
```diff
@@ -88,7 +89,15 @@ def do_flash(self, **kwargs):
+            download_args += [self.hex_file]
+        else:
+            if self.elf_name is None:
+                raise ValueError(f'Cannot flash; elf file required.')
```
Sorry to disagree with @pdgendt but the nesting does not seem needed. This is also has more consistent negations and looks more like a case/switch (which could also be an option?

```python
  if self.hex_file is not None:
     ...
  elif self.elf_name is not None:
     ...
  else:
    raise ValueError(f'Cannot flash; elf file required.')
```

No big deal.

marc-hb reviewed (commented):
More important than the codestyle: can both options be passed at the same time and if yes then do you really want HEX to have precedence? Should that be documented somewhere?

(Keep in mind this sort of script is not just used by people but also by higher level scripts and automation)

AndreasErnst-EH committed:
runners: probe-rs: add support to flash hex file

The probe-rs runner always flashed the elf file.
With this commit support for the --hex-file parameter is added.
This can be helpful when wanting to flash a signed binary instead
of the unsigned elf file.

Signed-off-by: Andreas Ernst <andreas.ernst@codewrights.de>

marc-hb reviewed (commented):
So can you pass both options and if yes then what happens?

pdgendt reviewed (changes requested):
I'm going to NAK this PR to favour https://github.com/zephyrproject-rtos/zephyr/pull/93670 instead.
